docs/cluster/networking.md

```diff
@@ -0,0 +1,142 @@
+# Networking in Kubernetes
+
+## Docker model
+
+Before discussing the Kubernetes approach to networking, it is worthwhile to
+review the "normal" way that networking works with Docker.  By default, Docker
+uses host-private networking.  It creates a virtual bridge, called `docker0` by
+default, and allocates an [RFC1918](https://tools.ietf.org/html/rfc1918) range
+for that bridge.  Each container that Docker creates is allocated a virtual
+ethernet device (called `veth`).  The other end of the veth is connected to the
+bridge.  The in-container interface is given an IP address from the bridge's
+address range.
+
+The result is that Docker containers can talk to other containers only if they
+are on the same machine (and thus the same virtual bridge).  Containers on
+different machines can not reach each other - in fact they may end up with the
+exact same network ranges and IP addresses.
+
+In order for Docker containers to communicate across nodes, they must be
+allocated ports on the machine's own IP address, which are then forwarded or
+proxied to the containers.  This obviously means that containers must either
+coordinate which ports they use very carefully or else be allocated ports
+dynamically.
+
+## Kubernetes model
+
+Coordinating ports across multiple developers is very difficult and runs against
+the general mindset of Kubernetes.  Dynamic port allocation brings a lot of
+complications to the system - every application has to take ports as flags,
+the API servers have to know how to insert dynamic port numbers into
+configuration blocks, services have to know how to find each other, etc.  Rather
+than deal with this, Kubernetes takes a different approach.
+
+The fundamental requirements that Kubernetes makes about networking are:
+   * all containers can communicate with all other containers without NAT
```

Comment:
Until we have that kind of security, anything you put here will just be confusing IMO. But once we have it, you could just add something at the end like "(except containers that are explicitly prohibited from communicating due to security policy)"